### ISSUE TITLE
Rotation transformation

### ISSUE THREAD

**Issue by noeagles** [pytorch/vision#566]
I noticed that RandomRotation  range only from (min, max); however it is more realistic to rotate from the set (angle1, angle2, angle3) randomly?

**Comment by fmassa** [comment]
Could you be a bit more clear on what you mean by those three angles?

**Comment by fmassa** [comment]
If I understand correctly, you are saying that we should propose the rotations to be randomly sampled from a discrete subset, instead of having them being sampled from a continuous interval, is that right?

If that's the case, I'd say that having it being sampled from an interval is more generic, and should probably be kept.

**Comment by sotte** [comment]
I agree with @fmassa. That being said, @noeagles, you could use the [functional transforms](https://pytorch.org/docs/stable/torchvision/transforms.html#module-torchvision.transforms.functional) and build that functionality yourself.
```python
import torchvision.transforms.functional as TF

angle = random.sample([-30, -15, 0, 15, 30])
img = TF.rotate(img, angle)
```

**Comment by VesninEN** [comment]
How it can be used inside the  transforms.Compose([...])?

**Comment by sotte** [comment]
Something like this would work:

```python
import torchvision.transforms.functional as TF
import random
from typing import Sequence

class MyRotateTransform:
    def __init__(self, angles: Sequence[int]):
        self.angles = angles

    def __call__(self, x):
        angle = random.choice([-30, -15, 0, 15, 30])
        return TF.rotate(x, angle)
```

**Comment by fmassa** [comment]
@sotte do you think this snippet could be added somewhere in torchvision? Maybe in the documentation of `rotate`?

**Comment by sotte** [comment]
@fmassa Um, if you think it's helpful, sure :) 

If we add this we should maybe just document the general procedure how you can use `TF` to build your own transform classes (and use the example from above).

I guess I could add it to https://pytorch.org/docs/stable/torchvision/transforms.html#functional-transforms.

I can create a PR (let me know if want something specific), but you can also just add the snippet. Both is fine with me.

**Comment by fmassa** [comment]
I think a note in the documentation of the functional interface, explaining that it gives full flexibility sounds like best thing to do, with an example for a particular case, like the rotation that you proposed just above

**Comment by sotte** [comment]
I'll send the PR later today.

**Comment by crypdick** [comment]
@fmassa I have a use-case which prohibits the use of  "generic interval" rotations because it destroys information I need in the corners. In particular, I can only use order-4 symmetry transforms (image attached). Using only reflections (RandomHorizontalFlip and RandomVerticalFlip) I can only get 4 of the 8 possible transforms.


![](https://upload.wikimedia.org/wikipedia/commons/thumb/c/cc/Dih4_cycle_graph.svg/800px-Dih4_cycle_graph.svg.png)
wiki/File:Dih4_cycle_graph.svg

These order-4 symmetry transforms are ubiquitous, for instance, in computer vision for satellite imagery research. I'm surprised that it isn't a built-in.

**Comment by fmassa** [comment]
@crypdick that is a valid point.

@vfdev-5 thoughts on this?

**Comment by vfdev-5** [comment]
@crypdick yes, rotations by 90 degrees as you suggest are definitely make sense for satellite, medical imagery and as data augmentations in torchvision. IMO, the reason why this is not yet built-in in torchvision, probably, due to "photography" type of imagery intended to be processed (like ImageNet) where upside-down rotation does not make much sense...

@fmassa I agree that this can be an enhancement for transforms and functional operations.

**Comment by YannDubs** [comment]
I agree that having the D_4 group augmentation as shown by @crypdick is very useful for many application ( biological, chemical and satellite data comes to mind).

It is very easy to get it though:
```python
from torchvision.transforms import (
    Compose,
    RandomApply,
    RandomHorizontalFlip,
    RandomRotation,
    RandomVerticalFlip,
)

transform = transforms.Compose(
            [
                RandomVerticalFlip(p=0.5),
                RandomVerticalFlip(p=0.5),
                RandomApply([RandomRotation(90, 90)], p=0.5),
            ]
        )
```

But a shorthand might be useful.
